File: ЛР2.cpp
// file: data_structures.cpp
#include <bits/stdc++.h>
using namespace std;

/*
1) Мультисписок (многоуровневый связный список)
Node имеет next и child; функция flatten превращает в одноуровневый список (DFS).
*/
struct MNode {
    int val;
    MNode* next;
    MNode* child;
    MNode(int v): val(v), next(nullptr), child(nullptr) {}
};

MNode* flatten(MNode* head) {
    if(!head) return head;
    MNode* cur = head;
    stack<MNode*> st;
    while(cur) {
        if(cur->child) {
            if(cur->next) st.push(cur->next);
            cur->next = cur->child;
            cur->child = nullptr;
        }
        if(!cur->next && !st.empty()) {
            cur->next = st.top(); st.pop();
        }
        cur = cur->next;
    }
    return head;
}

// helper to print
void printList(MNode* head) {
    while(head) {
        cout << head->val << " ";
        head = head->next;
    }
    cout << "\n";
}

/*
2) Deque, Queue, Priority Queue (primitive) and Priority Queue (custom struct)
*/
void cpp_stl_examples() {
    // Deque (двусторонняя)
    deque<int> dq;
    dq.push_back(2);
    dq.push_front(1);
    dq.push_back(3);
    cout << "Deque front: " << dq.front() << ", back: " << dq.back() << "\n";
    dq.pop_front(); // remove 1

    // Queue (FIFO)
    queue<int> q;
    q.push(10);
    q.push(20);
    cout << "Queue front: " << q.front() << "\n";
    q.pop();

    // Priority queue (max-heap, primitive)
    priority_queue<int> pqmax;
    pqmax.push(5); pqmax.push(1); pqmax.push(7);
    cout << "PQ max top: " << pqmax.top() << "\n"; // 7

    // Priority queue (min-heap) using greater<>
    priority_queue<int, vector<int>, greater<int>> pqmin;
    pqmin.push(5); pqmin.push(1); pqmin.push(7);
    cout << "PQ min top: " << pqmin.top() << "\n"; // 1

    // Priority queue with custom struct
    struct Task {
        int priority;
        string name;
    };
    struct Cmp {
        bool operator()(Task const& a, Task const& b) const {
            return a.priority < b.priority; // bigger priority first (max-heap)
        }
    };
    priority_queue<Task, vector<Task>, Cmp> task_pq;
    task_pq.push({2, "low"});
    task_pq.push({10, "high"});
    cout << "Top task: " << task_pq.top().name << " (p=" << task_pq.top().priority << ")\n";
}

int main() {
    // --- Демонстрация мультисписка ---
    MNode* n1 = new MNode(1);
    MNode* n2 = new MNode(2);
    MNode* n3 = new MNode(3);
    MNode* n4 = new MNode(4);
    MNode* n5 = new MNode(5);

    // level 0: 1 -> 2 -> 3
    n1->next = n2; n2->next = n3;
    // child of 2: 4 -> 5
    n2->child = n4; n4->next = n5;

    cout << "Before flatten: ";
    // print only top-level
    MNode* tmp = n1;
    while(tmp) { cout << tmp->val << " "; tmp = tmp->next; }
    cout << "\n";

    MNode* head = flatten(n1);
    cout << "After flatten: "; printList(head);

    // --- STL examples ---
    cpp_stl_examples();
    return 0;
}

File: ЛР2.java
// file: DataStructures.java
import java.util.*;
import java.util.concurrent.PriorityBlockingQueue;

public class DataStructures {

    // 1) Мультисписок (Node с next и child) и flatten
    static class MNode {
        int val;
        MNode next;
        MNode child;
        MNode(int v) { val = v; next = null; child = null; }
    }

    static MNode flatten(MNode head) {
        if (head == null) return head;
        Deque<MNode> stack = new ArrayDeque<>();
        MNode cur = head;
        while (cur != null) {
            if (cur.child != null) {
                if (cur.next != null) stack.push(cur.next);
                cur.next = cur.child;
                cur.child = null;
            }
            if (cur.next == null && !stack.isEmpty()) {
                cur.next = stack.pop();
            }
            cur = cur.next;
        }
        return head;
    }

    static void printList(MNode head) {
        while (head != null) {
            System.out.print(head.val + " ");
            head = head.next;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        // Deque
        ArrayDeque<Integer> dq = new ArrayDeque<>();
        dq.addLast(2); dq.addFirst(1); dq.addLast(3);
        System.out.println("Deque front: " + dq.peekFirst() + ", back: " + dq.peekLast());
        dq.removeFirst();

        // Queue (FIFO) using LinkedList
        Queue<Integer> q = new LinkedList<>();
        q.add(10); q.add(20);
        System.out.println("Queue front: " + q.peek());
        q.remove();

        // PriorityQueue primitive (default in Java — min-heap)
        PriorityQueue<Integer> pqMin = new PriorityQueue<>();
        pqMin.add(5); pqMin.add(1); pqMin.add(7);
        System.out.println("PQ min top: " + pqMin.peek());

        // Max-heap using Comparator.reverseOrder()
        PriorityQueue<Integer> pqMax = new PriorityQueue<>(Comparator.reverseOrder());
        pqMax.add(5); pqMax.add(1); pqMax.add(7);
        System.out.println("PQ max top: " + pqMax.peek());

        // PriorityQueue with custom class
        class Task {
            int priority;
            String name;
            Task(int p, String n) { priority = p; name = n; }
        }
        // Comparator: larger priority first
        PriorityQueue<Task> taskPQ = new PriorityQueue<>( (a,b) -> Integer.compare(b.priority, a.priority) );
        taskPQ.add(new Task(2, "low"));
        taskPQ.add(new Task(10, "high"));
        Task top = taskPQ.peek();
        System.out.println("Top task: " + top.name + " (p=" + top.priority + ")");

        // Demo multi-list
        MNode n1 = new MNode(1);
        MNode n2 = new MNode(2);
        MNode n3 = new MNode(3);
        MNode n4 = new MNode(4);
        MNode n5 = new MNode(5);
        n1.next = n2; n2.next = n3;
        n2.child = n4; n4.next = n5;
        System.out.print("Before flatten: ");
        MNode t = n1;
        while (t != null) { System.out.print(t.val + " "); t = t.next; }
        System.out.println();
        flatten(n1);
        System.out.print("After flatten: "); printList(n1);
    }
}
